[VectorCombine] Add foldShuffleToIdentity

davemgreen opened the pull request:
This patch adds a basic version of a combine that attempts to remove shuffles that when combined simplify away to an identity shuffle. For example:
  %ab = shufflevector <8 x half> %a, <8 x half> poison, <4 x i32> <i32 3, i32 2, i32 1, i32 0>
  %at = shufflevector <8 x half> %a, <8 x half> poison, <4 x i32> <i32 7, i32 6, i32 5, i32 4>
  %abt = fneg <4 x half> %at
  %abb = fneg <4 x half> %ab
  %r = shufflevector <4 x half> %abt, <4 x half> %abb, <8 x i32> <i32 7, i32 6, i32 5, i32 4, i32 3, i32 2, i32 1, i32 0>
By looking through the shuffles and fneg, it can be simplified to:
  %r = fneg <8 x half> %a

The code tracks each lane starting from the original shuffle, keeping a track of a vector of {src, idx}. As we propagate up through the instructions we will either look through intermediate instructions (binops and unops) or see a collections of lanes that all have the same src and incrementing idx (an identity). We can also see a single value with identical lanes, which we can treat like a splat.

Only the basic version is added here, handling identities, splats, binops and unops. In follow-up patches other instructions can be added such as constants, intrinsics, cmp/sel and zext/sext/trunc.

RKSimon commented:
Funnily enough, I have started work on a VectorCombine::foldShuffleOfShuffles fold that will sort of do this (and catch many of these test cases):

 `shuffle(shuffle(x,u,m1),shuffle(y,u,m2)) -> shuffle(x,y,m3)`

But it doesn't recurse, so every fold has to be cost beneficial - are you seeing cases where we need the recursion capability?

RKSimon commented:
Yes - we've never come up with a good way to accumulate candidate folds and only commit them if/when it becomes cost effective.

Please don't think I want to just take #88743 + #88899 instead of this, I think they're both needed, I just want to make sure we don't end up with a heavy fold constantly doing too much of the combining.

davemgreen commented:
The trivial cases were just added for testing - many of them are already handled by instcombine. Some of those combines in instcombine are the ones causing the problem, as they already move shuffles through binops without considering the costs and breaking the canonical patterns for interleaving loads.

We were previously trying to fix that in the InterleavedAccess (through replaceBinOpShuffles), but that starts to become impossible as more and more folds happen in the mid-end.

That is separate from this patch though, which shouldn't create the same problems of breaking interleaving patterns without replacing them with identity shuffles, which should always be an improvement.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -1547,6 +1548,148 @@ bool VectorCombine::foldShuffleOfCastops(Instruction &I) {
+          cast<FixedVectorType>(SV->getOperand(0)->getType())->getNumElements();
+      int M = SV->getMaskValue(Lane);
+      if (M < 0)
+        return {nullptr, -1};
```
```suggestion
        return {nullptr, PoisonMaskElem};
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -1547,6 +1548,148 @@ bool VectorCombine::foldShuffleOfCastops(Instruction &I) {
+  };
+
+  auto GenerateInstLaneVectorFromOperand =
+      [&LookThroughShuffles](const SmallVector<InstLane> &Item, int Op) {
```
```suggestion
      [&LookThroughShuffles](ArrayRef<InstLane> Item, int Op) {
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -1547,6 +1548,148 @@ bool VectorCombine::foldShuffleOfCastops(Instruction &I) {
+        for (InstLane V : Item) {
+          NItem.emplace_back(
+              !V.first
+                  ? InstLane{nullptr, -1}
```
```suggestion
                  ? InstLane{nullptr, PoisonMaskElem}
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -1547,6 +1548,148 @@ bool VectorCombine::foldShuffleOfCastops(Instruction &I) {
+    if (!all_of(drop_begin(Item), [&](InstLane IL) {
+          if (!IL.first)
+            return true;
+          if (isa<Instruction>(IL.first) &&
```
Use dyn_cast

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -1547,6 +1548,148 @@ bool VectorCombine::foldShuffleOfCastops(Instruction &I) {
+
+  // If we got this far, we know the shuffles are superfluous and can be
+  // removed. Scan through again and generate the new tree of instructions.
+  std::function<Value *(const SmallVector<InstLane> &)> generate =
```
```suggestion
  auto Generate =
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -1547,6 +1548,148 @@ bool VectorCombine::foldShuffleOfCastops(Instruction &I) {
+  // If we got this far, we know the shuffles are superfluous and can be
+  // removed. Scan through again and generate the new tree of instructions.
+  std::function<Value *(const SmallVector<InstLane> &)> generate =
+      [&](const SmallVector<InstLane> &Item) -> Value * {
```
```suggestion
      [&](ArrayRef<InstLane> Item) -> Value * {
```

davemgreen committed:
Address comments and fix failing test

davemgreen committed:
[VectorCombine] Add foldShuffleToIdentity

This patch adds a basic version of a combine that attempts to fold away
shuffles that when combines simplify away to an identity shuffle.
For example:
  %ab = shufflevector <8 x half> %a, <8 x half> poison, <4 x i32> <i32 3, i32 2, i32 1, i32 0>
  %at = shufflevector <8 x half> %a, <8 x half> poison, <4 x i32> <i32 7, i32 6, i32 5, i32 4>
  %abt = fneg <4 x half> %at
  %abb = fneg <4 x half> %ab
  %r = shufflevector <4 x half> %abt, <4 x half> %abb, <8 x i32> <i32 7, i32 6, i32 5, i32 4, i32 3, i32 2, i32 1, i32 0>
By looking through the shuffles, it can be simplified to:
  %r = fneg <8 x half> %a

The code tracks each lane starting from the original shuffle, keeping a track
of a vector of {src, idx}. As we propagate up through the instructions we will
either look through intermediate instructions (binops and unops) or see a
collections of lanes that all have the same src and incrementing idx (an
identity). We can also see a single value with identical lanes, which we can
treat like a splat.

Only the basic version is added here, handling identites, splats, binops and
unops. In follow-up patches other instructions can be added such as constants,
intrinsics, cmp/sel and zext/sext/trunc.

davemgreen committed:
Rebase and update phase-ordering tests

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -1547,6 +1548,147 @@ bool VectorCombine::foldShuffleOfCastops(Instruction &I) {
+        return NItem;
+      };
+
+  SmallVector<InstLane> Start;
+  for (unsigned M = 0; M < Ty->getNumElements(); ++M)
+    Start.push_back(LookThroughShuffles(&I, M));
```
```suggestion
  SmallVector<InstLane> Start(Ty->getNumElements());
  for (unsigned M = 0, E = Ty->getNumElements(); M < E; ++M)
    Start[M]= LookThroughShuffles(&I, M);
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -1547,6 +1548,147 @@ bool VectorCombine::foldShuffleOfCastops(Instruction &I) {
+    if (!all_of(drop_begin(Item), [&](InstLane IL) {
+          if (!IL.first)
+            return true;
+          if (auto *I = dyn_cast<Instruction>(IL.first); I && !I->hasOneUse())
+            return false;
+          return IL.first->getValueID() == Item[0].first->getValueID() &&
+                 (!isa<IntrinsicInst>(IL.first) ||
+                  cast<IntrinsicInst>(IL.first)->getIntrinsicID() ==
+                      cast<IntrinsicInst>(Item[0].first)->getIntrinsicID());
```
```suggestion
          auto *I = dyn_cast<Instruction>(IL.first);
          if (I && !I->hasOneUse())
            return false;
          return IL.first->getValueID() == Item[0].first->getValueID() &&
                 (!I || I->getIntrinsicID() ==
                      cast<IntrinsicInst>(Item[0].first)->getIntrinsicID());
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -1547,6 +1548,147 @@ bool VectorCombine::foldShuffleOfCastops(Instruction &I) {
+
+  // If we got this far, we know the shuffles are superfluous and can be
+  // removed. Scan through again and generate the new tree of instructions.
+  std::function<Value *(ArrayRef<InstLane>)> generate =
+      [&](ArrayRef<InstLane> Item) -> Value * {
```
```suggestion
  auto Generate =
      [&](ArrayRef<InstLane> Item) -> Value * {
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -1547,6 +1548,147 @@ bool VectorCombine::foldShuffleOfCastops(Instruction &I) {
+                                      E.value().second == (int)E.index());
+        })) {
+      return Item[0].first;
+    } else if (SplatLeafs.contains(Item[0].first)) {
```
```suggestion
    if (SplatLeafs.contains(Item[0].first)) {
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -1547,6 +1548,147 @@ bool VectorCombine::foldShuffleOfCastops(Instruction &I) {
+    }
+
+    auto *I = cast<Instruction>(Item[0].first);
+    SmallVector<Value *> Ops;
+    unsigned E = I->getNumOperands();
+    for (unsigned Idx = 0; Idx < E; Idx++)
+      Ops.push_back(generate(GenerateInstLaneVectorFromOperand(Item, Idx)));
```
```suggestion
    SmallVector<Value *> Ops(I->getNumOperands());
    for (unsigned Idx = 0, E = I->getNumOperands(); Idx < E; Idx++)
      Ops[Idx] = Generate(GenerateInstLaneVectorFromOperand(Item, Idx));
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -1547,6 +1548,147 @@ bool VectorCombine::foldShuffleOfCastops(Instruction &I) {
+    if (auto BI = dyn_cast<BinaryOperator>(I))
+      return Builder.CreateBinOp((Instruction::BinaryOps)BI->getOpcode(),
+                                 Ops[0], Ops[1]);
+    if (auto UI = dyn_cast<UnaryOperator>(I))
+      return Builder.CreateUnOp((Instruction::UnaryOps)UI->getOpcode(), Ops[0]);
+    llvm_unreachable("Unhandled instruction in generate");
```
```suggestion
    assert(isa<UnaryOperator>(I) && "Expected either Unary or Binary Op.");
    return Builder.CreateUnOp((Instruction::UnaryOps)I->getOpcode(), Ops[0]);
```

davemgreen committed:
Address more comments

RKSimon commented on llvm/test/Transforms/VectorCombine/AArch64/shuffletoidentity.ll:
```diff
@@ -192,11 +164,9 @@ define <8 x half> @splatandidentity(<8 x half> %a, <8 x half> %b) {
-; CHECK-NEXT:    [[R:%.*]] = shufflevector <4 x half> [[AB1]], <4 x half> [[AB2]], <8 x i32> <i32 7, i32 6, i32 5, i32 4, i32 3, i32 2, i32 1, i32 0>
+; CHECK-NEXT:    [[TMP1:%.*]] = shufflevector <8 x half> [[B:%.*]], <8 x half> poison, <8 x i32> zeroinitializer
+; CHECK-NEXT:    [[TMP2:%.*]] = shufflevector <8 x half> [[A:%.*]], <8 x half> poison, <8 x i32> zeroinitializer
+; CHECK-NEXT:    [[R:%.*]] = fadd <8 x half> [[TMP2]], [[TMP1]]
```
This looks like we might need a foldBinopOfShuffles as well......

RKSimon reviewed (commented):
Do you intend to support intrinsics at some point? I notice you have a abs test

davemgreen commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -1667,6 +1668,149 @@ bool VectorCombine::foldShuffleOfShuffles(Instruction &I) {
+    if (all_of(drop_begin(Item), [&](InstLane &IL) {
+          return !IL.first ||
+                 (IL.first == Item[0].first && IL.second == Item[0].second);
+        })) {
```
We don't currently handle undef zero'th lanes due to the complexities it causes throughout the combine. It is the `if (!Item[0].first)` above line that checks if the first element is undef.

davemgreen reviewed (commented):
> Do you intend to support intrinsics at some point? I notice you have a abs test

Yep, those and a few other operations. I wanted to make sure the basics were working before adding the others but they are hopefully relatively straight forward.

davemgreen committed:
Rebase over other changes and fixup tests.

davemgreen commented:
Rebased and ping. Thanks

RKSimon reviewed (approved):
Final question - do we need handling for DivRem instructions to make sure we don't introduce poison mask elements? see 4cc9c6d98dfef90d1ffa69977d13ffa2894a10f7

davemgreen commented:
> Final question - do we need handling for DivRem instructions to make sure we don't introduce poison mask elements? see [4cc9c6d](https://github.com/llvm/llvm-project/commit/4cc9c6d98dfef90d1ffa69977d13ffa2894a10f7)

Thanks - I hadn't considered div/rem as they don't as often come up for vectors. I can exclude them for now if that sounds OK, but as far as I understand they might be fine as we end up removing the shuffles so do not have poison lanes left over.
https://alive2.llvm.org/ce/z/NAyQty

davemgreen committed:
Exclude div/rem

RKSimon reviewed (approved):
LGTM - cheers

dyung commented:
Hi @davemgreen, we are seeing an assertion failure in a few internal tests which I bisected back to your change. I have filed #91078 for the issue, can you take a look?

davemgreen commented:
Thanks for the heads up - I'll take a look.

davemgreen commented:
Hi - yeah I'll take a look. Thanks for letting me know.
